### ISSUE TITLE
shallow copy / view for some blocks in a BlockVector (interface to Anasazi eigensolvers)

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#5185]
I am trying to figure out how to add support for shallow copies/views to BlockVectors used as multi-vectors (i.e. all blocks have the same distribution and size), 

That's something Anasazi (Trilinos eigenpackage) needs as a part of multi-vector traits. To quote Anasazi paper:

> The populated multivectors can be a deep copy, where the object contains the storage for the multivector entries, or a shallow copy, where the object has a view of another multivector’s storage. A shallow copy is useful when only a subset of the columns of a multivector is required for computation, which is a situation that commonly occurs during the generation of a subspace.

What I did so far is I switched  `class BlockVectorBase` from:
```
std::vector<VectorType> components;
```
to
```
std::vector<std::shared_ptr<VectorType> > components;
```

This shall allow getting a similar functionality to [Tpetra::MultiVector::subViewNonConst](https://trilinos.org/docs/dev/packages/tpetra/doc/html/classTpetra_1_1MultiVector.html#a288e1d94c10062baf4d8dec4b01ab3fd), that is  -- a view which can be modified. 

For example, to view two blocks, internally in `BlockVectorBase::view(BlockVectorBase &rhs, const std::vector<unsigned int> blocks)` something like this would happen:

```
this->components[0] = V.components[blocks[0]];
this->components[1] = V.components[blocks[1]];
```

The tricky part is a constant view: given `const BlockVector &block_vector`, I would need to get a const view to some blocks, like [TPetra::MultiVector::subView](https://trilinos.org/docs/dev/packages/tpetra/doc/html/classTpetra_1_1MultiVector.html#a84904aa971f75bcd0dc830c8b696b7c8).

Here one can't directly play with shared pointers as the input vector is `const`. I am scratching my head how to address this. I would appreciate any suggestions / hints.

p.s. the idea is to have adaptors for Anasazi to use LA::d::BlockVector as multi-vectors.

**Comment by bangerth** [comment]
I'm not excited about this. It changes the model we have for vectors.

Couldn't you write an entirely separate class that does what you want -- say, just like `ArrayView` gives a view, you'd create a class `MultiArrayView` that provides the view you need?

**Comment by davydden** [comment]
> I'm not excited about this. It changes the model we have for vectors.

well, but this is cosmetic and purely internal to users / applications. Wha is your concern?

> Couldn't you write an entirely separate class that does what you want

i don't think this is the interface which is required. Those views/shallow copies should return objects of the same type. So `BlockVector` in -> `BlockVector` out.

**Comment by davydden** [comment]
p.s. there could be indeed a better way to achieve this, that's the reason for a discussion.

**Comment by bangerth** [comment]
On 10/02/2017 04:24 PM, Denis Davydov wrote:
>     I'm not excited about this. It changes the model we have for vectors.
> 
> well, but this is cosmetic and purely internal to users / applications. Wha is 
> your concern?

That suddenly BlockVector works differently from Vector and, in fact, every 
other vector type we have -- all of which *own* their memory outright, rather 
than sharing it. I don't think it's the right approach to revisit this design 
decision to accommodate a single use case. There is a benefit in using common 
design decisions across classes, even if it's not visible to users -- that 
code still has to be maintained, after all.

**Comment by davydden** [comment]
> I don't think it's the right approach to revisit this design
decision to accommodate a single use case. 

But you also don’t suggest an alternative...

**Comment by bangerth** [comment]
Well, I did suggest a `MultiArrayView` :-)

I think I haven't understood yet what exactly it is you want to do. Can you outline what use case you have? Then we can think of how to design and implement that, but start with what you want, not how you want it.

**Comment by davydden** [comment]
@bangerth 

> start with what you want, not how you want it.

I kind of tried to outline it in the OP. So let me try again:

I want to add native support to Anasazi eigensolvers form Trilinos. To do that [1]:

> All that is required is the template instantiation of the trait classes, MultiVecTraits and OperatorTraits, for the user-defined multivector and operator, respectively.

[MultiVecTraits](https://trilinos.org/docs/dev/packages/anasazi/doc/html/classAnasazi_1_1MultiVecTraits.html#ae006d07804e9786059f3b6481543cf97) and [OperatorTraits](https://trilinos.org/docs/dev/packages/anasazi/doc/html/classAnasazi_1_1OperatorTraits.html). Here is a native example of Trilinos [TpetraAdapters](https://github.com/trilinos/Trilinos/blob/master/packages/anasazi/tpetra/src/AnasaziTpetraAdapter.hpp). Roughly, one needs to be able to : 
1. clone given multt-vector or some of its blocks (**easy**)
2. create const and non-const views to some columns in the given block vector (that's what we discuss, actually) (**PROBLEMATIC**)
3. some operations with dense matrices, like `B1 = a B2 S + B3` where `S` is dense matrix and `B1`,`B2`,`B3` are block-vectors (**easy**)
4. initialization with random numbers (**easy**)
5. apply operator to a multi-vector (say MatrixFree operator on Vectors applied to each block in a block vector). The underlying operator can be a matrix, preconditioner or spectral transformation (**should be easy with our linear operator concept**).

Again, to quote [1]:

> The populated multivectors can be a deep copy, where the object contains the storage for the multivector entries, or a shallow copy, where the object has a view of another multivector’s storage. A shallow copy is useful when only a subset of the columns of a multivector is required for computation, which is a situation that commonly occurs during the generation of a subspace.

Obviously I have not gone through the whole process yet, but I would say the fact that there are views/shallow copies automatically implies that objects need to share resources. But I also don't know how Trilinos implement this in Epetra or Tpetra.

What is Anasazi as I see it:

1. collection of algorithms for the numerical solution of large-scale eigenvalue problems
2. independent of the choice of the underlying linear algebra primitives by using abstract interfaces. Matrices and vectors are opaque objects, user needs to provide a set of elementary operations.
3. one can use their building-blocks (i.e. orthogonalization routines) to implement new solvers.

From [1]

> Our ap- proach is algorithm oriented [Musser and Stepanov 1994], in that requirements for efficient implementation of the necessary algorithms were considered first. This was followed by a formulation of the software abstractions capable of im- plementing these algorithms, and their constituent mechanisms, in sufficiently diverse ways. The result was a collection of implementations that are both ef- ficient and flexible

which is a raison d’etre for the adapters.

Long story short -- it's an object-oriented eigensolvers for standard and generalized problems and is a good competitor to the aging Arpack. It is independent of other Trilinos packages and only needs: Trilinos Teuchos for RCP (a reference-counting smart pointer), ParameterList and Blas/Lapack (for dense LA).

p.s. _multivector_ == dense rectangular matrix. Essentially our _block_ vector where each block has the same size / partition / etc.

p.p.s. more details are in the  section "4.1 The Anasazi Operator/Multivector Interface" of [1]

DISCLAIMER: I always think of an eigenproblem to be solved as a collection of non-block vectors. So if the operators themselves are block operators, then this requires a multi-vector where each vector is a block vector.

[1] Becker et al, 2009, Anasazi Software for the Numerical Solution of Large-Scale Eigenvalue Problems

**Comment by guidokanschat** [comment]
I find this an interesting approach.

First, I would really appreciate a well-designed interface to a modern eigensolver suite. The more I learn about eigenvalues, the less satisfied I am with the current arpack support.

Second, I see a second use for this approach. We always had the problem that we cannot build a solver for more complex physics based on an existing solver for less complex problems.

Say, you have a working Navier-Stokes solver working with block matrices and block vectors, and you want to do Boussinesq. You basically end up reimplementing the Stokes solver in a system with additional temperature, since you have one system with a two-component `BlockVector` $(u,p)$ and the other with three components $(u,p,T)$. We have never got around to implement nested block vectors for such problems.

The structure that Denis proposes renders this nested block vector obsolete. You can simply compose a reduced block vector from the blocks of the original one **without** copying the entries. To me this looks like a considerable improvement of `BlockVector`.

Additionally, I would like to comment on the "works differently" part: The beauty of C++ lies in the fact that the apparent use of an object (here, access to entries of a block vector as well as to the blocks) is independent from the actual implementation, such that the implementation can be improved when the need arises. I am not a friend of the "it's always been so" reasoning.

Ownership of the data is not an exported feature of any vector class, and it is a historical accident. In fact, the arpack interface would be a lot less botched up if vector did not own its data. While I am not seriously suggesting a vector with data in the middle of a big chunk of FORTRAN data, I am finding more and more reasons why vector should be able to use existing data, such that we can avoid copying. Therefore, using a shared pointer with an offset in `Vector` does not seem outrageous to me.

Naturally, all of this can be done using a `View`. But then, all functions that used to work on `Vector` have to work on `View`. Which means, the `View` is the new `Vector`, and the vector is just a glorified `std::vector` storing the data.

**Comment by bangerth** [comment]
@davydden -- I looked through the links you provide above and the way I see it is that what Anasazi wants is an object that *may* be a view to another vector, but that that is not a requirement.

I think what I object to is trying to fit the semantics of what Anasazi wants onto a class that we already have where I don't see that fitting. My preference would be to create a class `MultiVector` that stands alone and does exactly what you need. If you wrote it as 
```
  template <typename VectorType>
  class MultiVector { ... };
```
you even keep your options open as to what vector type it uses. Anasazi does not require a lot of operations, so its interface does not have to be nearly as extensive as that of our current vector classes.

I see two ways to implement it based on what Anasazi wants. The first would be for it to own its vectors. That could be a member variable `std::vector<VectorType> vectors;` or, if you want to efficiently resize it, maybe better `std::vector<std::unique_ptr<VectorType>> vectors;`. The alternative would be for it to use vectors owned elsewhere, and for example have a member `add_vector(VectorType &vec)` that adds another column to the multivector by referencing the vector just added. In that case, the vectors pointed to could, for example, be the individual blocks of a `BlockVector`, or just a collection of solution vectors all associated with one `DoFHandler`. In any case, I tend to think that having a new class that does exactly what you want is likely the easier option that trying to shoe-horn a different use onto `BlockVector`.

**Comment by bangerth** [comment]
@guidokanschat -- with the `LinearOperator` framework, it has become pretty easy to pick a subset of blocks of a block matrix and build an operator based on that -- say, for a Schur complement, or a block triangular operator. What we can't do, right now, is to select a subset of blocks of a block vector.

**Comment by davydden** [comment]
>  My preference would be to create a class MultiVector that stands alone and does exactly what you need.

But then it means that you need to extend all those other places in dealii which currently work by default with `BlockVectorBase` to this new `MultiVector` class, so quite some extra instantiations. Also some code duplications as well for simple operations like addition of two multi-vectors:
```
void BlockVectorBase<VectorType>::add(
const value_type a, 
const BlockVectorBase< VectorType >)
```

I don't really object this approach, just pointing out that it's not ideal either as it duplicates a lot of functionality already provided by`BlockVectorBase`.

**Comment by bangerth** [comment]
I think it comes down to the question where you want to use the new class. If it's only for Anasazi, then you don't need a big interface. Or what are your other plans?

Alternatively, you could implement a class that *interprets* the blocks of a block vector as the columns of a multivector. A bit like what you wanted to do at first, just in a separate class. In that case implementing most vector operations just boils down to forwarding to the corresponding `BlockVector` functions (or, if you wanted to be exceedingly lazy, deriving `private` and using `using` statements).

**Comment by davydden** [comment]
> I think it comes down to the question where you want to use the new class. If it's only for Anasazi, then you don't need a big interface. Or what are your other plans?

I currently approach problems with vector-space unknowns via GMG + Matrix-Free and BlockVectors. It would be good to use the same objects to have an alternative solution using eigensolvers (that's what I already have with Arpack as it manages its own memory for vectors anyway and only needs eigenvectors for output; I can simply use `std::vector<Vector*>` to get it working). 

> Alternatively, you could implement a class that interprets the blocks of a block vector as the columns of a multivector.
 
Not sure how this solves the problem of creating interface for Anasazi, which requires both deep and shallow copies. Say `BlockVector` owns its vectors and then `MultiVector` wrapper class internally only stores `std::vector<Vector*>` pointers to each blocks. This would be ideal for views/shallow copies, but now deep copies and destructors are problematic as one can't easily distinguish between a `MultiVector` which needs to clean-up things and the one which does not  as it stores pointers to blocks of `BlockVector`.

> what Anasazi wants is an object that may be a view to another vector, but that that is not a requirement.

I think it is a requirement. One needs to implement a shallow copy:

```
// Creates a new MV that shares the selected contents of mv (shallow copy).
static 
Teuchos::RCP<MV> 
Anasazi::MultiVecTraits< ScalarType, MV >::CloneViewNonConst (MV &mv, const std::vector<int> &index)
```
and a deep copy 
```
// Creates a new MV and copies the selected contents of mv into the new vector (deep copy).
static 
Teuchos::RCP<MV> 
Anasazi::MultiVecTraits< ScalarType, MV >:: CloneCopy (const MV &mv, const std::vector<int> &index)
```

In both cases you take the original object (to be used with  OperatorTraits) and return a pointer to the object of the same type.

-----

Can you elaborate what you actually don't like in sharing blocks between block-vectors?

**Comment by bangerth** [comment]
I think you already have the design then -- a class that has `std::shared_ptr` pointers to vectors. This allows both deep and shallow copies. This seems like a sensible implementation of the requirements.

> Can you elaborate what you actually don't like in sharing blocks between block-vectors?

I don't to having a class with these semantics. My objection is to have this in the `BlockVector` class because then `BlockVector` would behave different from `Vector`, `TrilinosWrappers::Vector`, `PETScWrappers::Vector`, etc. I put great value into uniformity of semantics because it makes programming so much easier: if you know how one class works, you know how they all work. If suddenly one class is different, then suddenly nothing you thought you know seems safe any more. 

There is also the issue that getting shared resources right is exceptionally difficult, both to implement and to define in its semantics. For example, if you allowed vectors to share resources, what is supposed to happen here:
```
void f(const Vector &v)
{
  Vector w = v;      // deep or shallow copy?
  ...
  w(1) = 42;  // does this affect 'v'? what about the fact that 'v' is a `const` vector?
  h();  // h is a non-const member function that can access 'v' and modifies it; does this affect 'w'?
```

This is the reason why, I believe, the Anasazi class wrappers don't define `operator=` or copy construction operators but carefully chooses function names that clearly indicate whether a function is supposed to create a copy, assign data, create a view, etc. That is good design. But we don't have that luxury any more -- for better or worse, we have `operator=` and allow `w = v`, and that leaves it ambiguous whether we mean `w.get_view_of(v)` or `w.copy_elements_from(v)` or anything else.

But as I said above, I'm very much in favor of a new class that (i) internally stores a collection of `Vector`s or one `BlockVector`, (ii) can explicitly manage the semantics that Anasazi needs, and (iii) implements its operations by just forwarding to the corresponding functions of the vectors or the block vector.

**Comment by davydden** [comment]
Just to clarify, because I think you mis-interprete my original intentions:

> My objection is to have this in the BlockVector class because then BlockVector would behave different from Vector, TrilinosWrappers::Vector, PETScWrappers::Vector, etc.

it will not behave differently, meaning that all public interface and its usage stays **as is**.

>  if you know how one class works, you know how they all work

nothing will change in the way you use block vectors. Code/tests which worked, should keep working.

> For example, if you allowed vectors to share resources, what is supposed to happen here:

exactly what used to happen before. I don't think I implied above that I will change any meaning/usage of any of the public functions of `BlockVectorBase`. The whole idea was **only to extend** `BlockVectorBase` to do something more, so if you don't use that functionality, things work as before.

> that leaves it ambiguous whether we mean w.get_view_of(v) or w.copy_elements_from(v) or anything else.

It's a matter of convention, if everything used to be a _deep_ copy, then only the extra functionality has anything to do with _shallow_ copies. But I agree that Anasazi have a better calling convention by distinguishing those two from the get go.

**Comment by bangerth** [comment]
I know you didn't mean to change the existing behavior but only to *enable* additional semantics.

But I still think it will be confusing that you can do fundamentally different things with block vectors than you can do with regular vectors. For example, we have code like this in a few places:
```
  void vmult (const Vector &src,
                    Vector &dst)
  {
    // ensure that source and destination are not the same:
    Assert (&dst != &src, ...);
```
The assertion used to guarantee that we can safely overwrite things in `dst` without affecting the `src` vector. But if you allow sharing of resources, this is suddenly no longer sufficient. So you say that you're not going to change any of the interfaces, but the semantics of objects change, and we would still have to rethink whether things that were correct in the past are still correct.

Some of the other examples were more of the kind "this opens a can of worms" if you allow sharing of resources. It's not that they change existing functionality, but that we would have to carefully document in what contexts one can or can not use shared resource vectors (i.e., some operations will require that a vector is/is not a view, or that a vector can not have any other views to it), and this will be quite difficult to figure out in some contexts. What's worse is that it only applies to `BlockVector` objects, but not to regular `Vector` objects.

**Comment by davydden** [comment]
thinking about it, the right way is probably first to workout multi-vectors on a level of a single vector (extend vectors in some way, by either adding new clalses or extending existing). Trilinos will have native support with EPetra multi-vectors. Note that some linear algebra packages may store multi-vectors in a specific way optimised to do sparse matrix multi vector (SpMMV) etc, so this beast is different from a simple collection of vectors. Once this is there, block vectors (meant to be used with block operators) can naturally combine those multi-vectors for each block. 

As the need is not pressing, and I am (ab)using block-vectors as multi-vectors for now, I am closing the issue for now.
